docs/resource/README.md

```diff
@@ -105,7 +105,50 @@ as specified in the [Resource SDK specification](https://github.com/open-telemet
+**[2]:** MUST be unique for each instance of the same `service.namespace,service.name` pair
+(in other words `service.namespace,service.name,service.instance.id` triplet MUST be globally unique).
+The ID helps to distinguish instances of the same service that exist at the same time
+(e.g. instances of a horizontally scaled service). It is preferable for the ID to be persistent
+and stay the same for the lifetime of the service instance, however it is acceptable that
+the ID is ephemeral and changes during important lifetime events for the service
+(e.g. service restarts).
+If the service has no inherent unique ID that can be used as the value of this attribute
+it is recommended to generate a random Version 1 or Version 4
+[RFC 4122](https://www.ietf.org/rfc/rfc4122.txt) UUID. Services aiming for reproducible UUIDs may also
+use Version 5. UUIDs are typically recommended, as we only need an opaque yet reproducible value for
+the purposes of identifying a service instance. Similar to what can be seen in the man page for the
+`/etc/machine-id` file, the underlying data, such as pod name and namespace should be treated as
+confidential by this algorithm, being the user's choice to expose it or not via another resource attribute.
+When a UUID v5 is generated, the UUID's namespace MUST be set to:
+`${telemetry.sdk.name}.${telemetry.sdk.language}.${service.name}`, so that the same service yields the
+same ID if the same value (`host.id`, `/etc/machine-id`, and so on) remains the same. It would still
+yield different results for different services on the same host.
+
+SDKs are required to use the following algorithm when generating `service.instance.id`:
+
+- If the user has provided a `service.instance.id`, via environment
+  variable, configuration or custom resource detection, this will
+  always be used and honored over generated IDs.
+- When any of the below combinations of resource attribute are provided, they should be used as the input
+  for generating a UUID v5. The values within each combination should be separated with dots:
+  * `container.id`
+  * `k8s.namespace.name`/`k8s.pod.name`/`k8s.container.name`
+  * `host.id`
+- When the SDK is running in an environment where a `/etc/machine-id`
+  (see [MACHINE-ID(5)](https://www.freedesktop.org/software/systemd/man/machine-id.html))
+  is available, the machine-id should be used as the input for generating a UUID v5.
+- When the SDK is running on a Windows environment and there's a reasonable way to read
+  registry keys for the SDK, the registry key
+  `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Cryptography\MachineGuid` can be used in a
+  similar way to Linux' machine-id above.
+- When no other source is available the SDK MUST generate a value using UUID v1 or v4.
```

Comment:
> When no other source is available 

Does this "other source" only refer to sources explicitly mentioned in the points above, or would SDKs be allowed to leverage _any_ other source that makes sense?

E. g., when running in Azure Service Fabric, [the service instance id](https://learn.microsoft.com/en-us/dotnet/api/system.fabric.statelessservicecontext.instanceid?view=azure-dotnet) would likely be a preferred choice. If we don't treat the list above as exhaustive that should be made more explicit.

Reply:
> Does this "other source" only refer to sources explicitly mentioned in the points above, or would SDKs be allowed to leverage any other source that makes sense?

SDKs are indeed allowed to use whatever makes sense. Is this ID only available for .NET? Or is there a common env var that would be available for all workloads?

Reply:
I wonder why we couldn't use more vague (and concise) wording like this:

> SDKs MUST determine the value `service.instance.id` in the following way:
>
> - If the user has provided a `service.instance.id`, via environment variable, configuration or custom resource detection, this is used and honored over generated IDs.
> - If a meaningful id can be generated that is unique for each instance of the same `service.namespace,service.name` pair, then this id is used.
> - Otherwise, the SDK generates a value using UUID v1 or v4.

It seems to be this should be enough to fulfill the requirement of having different ids for each running service instance. It also relieves us from maintaining a (quite complicated) list that will always be incomplete. If needed, there could still be recommendations about how a meaningful id can be generated (as part of this document or separately), but I think specifying, discussing, and implementing a detailed algorithm doesn't provide much benefit.

Reply:
That's true, but the value I see from specifying the behavior is that this is consistent across languages and wouldn't leave users wondering why one service written in X shows a consistent ID on a specific machine while another service written in Y and running on the same machine doesn't. We can split the list and precedence of those values into another doc, but those should exist in a place that maintainers can easily see.

Reply:
> SDKs are indeed allowed to use whatever makes sense.

Given that, you'll still not get fully consistent behavior across languages. However, you admittedly get more overlap.

Reply:
I believe those other use cases would be the fallbacks for specific situations. Given the same environments and conditions, two workloads would act the same. Before falling back to generating a random UUID, one language might still try another approach that is specific to it.